metals/src/main/scala/scala/meta/internal/metals/codeactions/FlatMapToForComprehensionCodeAction.scala

```diff
@@ -0,0 +1,423 @@
+package scala.meta.internal.metals.codeactions
+
+import scala.concurrent.ExecutionContext
+import scala.concurrent.Future
+
+import scala.meta.Term
+import scala.meta.Tree
+import scala.meta.inputs.Position
+import scala.meta.internal.metals.Buffers
+import scala.meta.internal.metals.CodeAction
+import scala.meta.internal.metals.MetalsEnrichments._
+import scala.meta.internal.parsing.Trees
+import scala.meta.io.AbsolutePath
+import scala.meta.pc.CancelToken
+
+import org.eclipse.lsp4j.CodeActionParams
+import org.eclipse.{lsp4j => l}
+
+class FlatMapToForComprehensionCodeAction(
+    trees: Trees,
+    buffers: Buffers
+) extends CodeAction {
+  override def kind: String = l.CodeActionKind.RefactorRewrite
+
+  override def contribute(params: CodeActionParams, token: CancelToken)(implicit
+      ec: ExecutionContext
+  ): Future[Seq[l.CodeAction]] = Future {
+    val uri = params.getTextDocument().getUri()
+
+    val path = uri.toAbsolutePath
+    val range = params.getRange()
+    val maybeTree =
+      if (range.getStart == range.getEnd)
+        trees
+          .findLastEnclosingAt[Tree](
+            path,
+            range.getStart(),
+            applyWithSingleFunction
+          )
+      else
+        None
+
+    val maybeCodeAction = for {
+      document <- buffers.get(path)
+      applyTree <- maybeTree
+      indentation = getIndentationForPosition(applyTree.pos)
+    } yield {
+      applyTree match {
+        case termApply: Term.Apply =>
+          maybeBuildToForYieldCodeActionWithApply(
+            path,
+            termApply,
+            document,
+            indentation
+          )
+
+        case termName: Term.Name
+            if termName.value == "flatMap" || termName.value == "map" =>
+          maybeBuildToForYieldCodeActionWithName(
+            path,
+            termName,
+            document,
+            indentation
+          )
+        case termIf: Term.If =>
+          maybeBuildToForYieldCodeActionWithIf(
+            path,
+            termIf,
+            document,
+            indentation
+          )
```

Comment:
It is only handled when `else` is `None` 
Also I want to cover for `emptyCollection`

Reply:
Which collections would you cover? There is way too much to cover `Seq()`, `Nil`, `List()`, `List.empty` etc.? There is dozens of possibilities that can work in this case and we will be unable to name them all.

Reply:
I think it would do to just cover the most common ones! In other cases, the if expression would not be handled! I think people defining their own None is very rare though. The developers who right such kind of code, would not use such advanced editor features which would significantly modify their code! Also defining ones own `None` in the code is a very bad practice!

Reply:
> I think it would do to just cover the most common ones! 

That's just it, it's not a very common pattern at all to write empty collections or None in an else branch. And requires semantic knowledge to make it work properly. Otherwise, we might break the user's code for very little gain.  

Even if it will work in most cases, we would be introducing a risky behavior for very little gain. 

For comprehensions can be translated from `filter`, `flatMap` and `map` chains, because they are syntactically equivalent. Doing anything else is out of scope for this issue and complicates it without reason.

Reply:
Ok, I would remove it!

Reply:
In Scala code, the type of the `then` branch is the same as the type of the `else` branch, isn’t it? I mean the `then` branch is going to be the nonEmpty collection of the `else`’s empty collection.

Reply:
It should if the code compiles, we might get non compiling code here that parses.